### ISSUE TITLE
What does the  'xhyk,bvk,bqk->bhvq'  mean???

### ISSUE THREAD

**Issue by jasscia18** [jnhwkim/ban-vqa#31]
What does this mean in the code,    logits = torch.einsum('xhyk,bvk,bqk->bhvq', (self.h_mat, v_, q_)) + self.h_bias
What does the  'xhyk,bvk,bqk->bhvq'  mean???

**Comment by jnhwkim** [comment]
Please refer to this [PyTorch Doc](https://pytorch.org/docs/stable/torch.html#torch.einsum). 
It says the `lower case letters (indices) to be associated with each dimension of the operands and result.`

**Comment by jasscia18** [comment]
I've seen this document, but I don't understand the 'xhyk,bvk,bqk->bhvq'  is how to calculate .
How do you compute these three matrices to get logits

**Comment by jasscia18** [comment]
I've seen this document, but I don't understand the 'xhyk,bvk,bqk->bhvq' is how to calculate .
How do you compute these three matrices to get logits




------------------ 原始邮件 ------------------
发件人: "Jin-Hwa Kim"<notifications@github.com>;
发送时间: 2019年7月10日(星期三) 下午4:24
收件人: "jnhwkim/ban-vqa"<ban-vqa@noreply.github.com>;
抄送: "　JAMIE "<2295410401@qq.com>;"Author"<author@noreply.github.com>;
主题: Re: [jnhwkim/ban-vqa] What does the  'xhyk,bvk,bqk->bhvq'  mean???(#31)




Please refer to this PyTorch Doc.
 It says the lower case letters (indices) to be associated with each dimension of the operands and result.

**Comment by jaesuny** [comment]
> I've seen this document, but I don't understand the 'xhyk,bvk,bqk->bhvq' is how to calculate .
> How do you compute these three matrices to get logits

`einsum` is a generalized version of matrix multiplication using the indexes of axes.
https://rockt.github.io/2018/04/30/einsum and https://obilaniu6266h16.wordpress.com/2016/02/04/einstein-summation-in-numpy/ may help.

The `einsum` you asked is the implementation of the low-rank bilinear pooling (eq 2 in the paper).
![image](https://user-images.githubusercontent.com/10977800/60957269-51135780-a33f-11e9-9c8e-66f447b1b20a.png)
`self.h_mat `, `q_` and `v_` correspond to P, U^T X, V^T Y.

For your information, `x` and `y` of  `'xhyk,bvk,bqk->bhvq'` is dummy indexes for compatibility of the previous version of BAN.

**Comment by jasscia18** [comment]
For flickr30k datasets  ,  In one batch, are all the boxes in the picture fused with all the entities of the corresponding sentences, or is this picture fused with all the entities of the sentences separately?




------------------ 原始邮件 ------------------
发件人: "Jaesun Park"<notifications@github.com>;
发送时间: 2019年7月10日(星期三) 下午5:32
收件人: "jnhwkim/ban-vqa"<ban-vqa@noreply.github.com>;
抄送: "　JAMIE "<2295410401@qq.com>;"Author"<author@noreply.github.com>;
主题: Re: [jnhwkim/ban-vqa] What does the  'xhyk,bvk,bqk->bhvq'  mean???(#31)



 
I've seen this document, but I don't understand the 'xhyk,bvk,bqk->bhvq' is how to calculate .
 How do you compute these three matrices to get logits
  
einsum is a generalized version of matrix multiplication using the indexes of axes.
 https://rockt.github.io/2018/04/30/einsum and https://obilaniu6266h16.wordpress.com/2016/02/04/einstein-summation-in-numpy/ may help.
 
The einsum you asked is the implementation of the low-rank bilinear pooling (eq 2 in the paper).
 
 self.h_mat , q_ and v_ correspond to P, U^T X, V^T Y.
 
For your information, x and y of  'xhyk,bvk,bqk->bhvq' is dummy indexes for compatibility of the previous version of BAN.

**Comment by jnhwkim** [comment]
@jasscia18 the features of boxes and the embeddings of entities.
